### PR TITLE
Adding fnv1 and fnv1a hash functions

### LINKED ISSUE #12500: Support FNV hash functions
https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function

There are different flavors of this hash function. We'd like to introduce
`bigint fnv64(binary)`
`bigint fnv32(binary)`

### PR DESCRIPTION
**Pull request by widagdos**
Fixes #12500.

```
== RELEASE NOTES ==

General Changes
* Added fnv1_32, fnv1_64, fnv1a_32, fnv1a_64,

```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    private static final long FNV_64_PRIME = 0x100000001b3L;
+
+    private static final int FNV_32_OFFSET_BASIS = 0x811c9dc5;
+    private static final int FNV_32_PRIME = 0x01000193;
+
+    private FnvHash()
+    {
+    }
+
+    static int fnv1_32(Slice data)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: Presto Java code follow camel convention (and don't usually use `_`). What about `fnv1Hash32`? ? Ditto for other methods.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java [hunk 2]
```diff
@@ -315,6 +315,38 @@ public static long crc32(@SqlType(StandardTypes.VARBINARY) Slice slice)
         return crc32.getValue();
     }
 
+    @Description("compute fnv-1 32 bit")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long fnv1_32(@SqlType(StandardTypes.VARBINARY) Slice slice)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: ditto for this method name :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java (hunk 1 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: can this be a public method?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import io.airlift.slice.Slice;
+
+// Taken from https://github.com/airlift/slice/blob/95d5876d3f4c57321767113e97416a42bc41749a/src/main/java/io/airlift/slice/FnvHash.java
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: What about the following?

```
/**
 * Reference implementation: https://tools.ietf.org/html/draft-eastlake-fnv-17#section-6
 * TODO: Move this utility class to airlift:slice
 */
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    {
+    }
+
+    static int fnv1_32(Slice data)
+    {
+        int hash = FNV_32_OFFSET_BASIS;
+
+        for (int i = 0; i < data.length(); ++i) {
+            int dataByte = data.getUnsignedByte(i);
+            hash *= FNV_32_PRIME;
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: Is is by deign that in this method, we first do multiplication and xor , while in `fnv1a_32` we first do xor then multiplication.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestVarbinaryFunctions.java [hunk 5]
```diff
@@ -364,6 +364,29 @@ public void testCrc32()
         assertFunction("crc32(to_utf8('ABCDEFGHIJKLM'))", BIGINT, 4223167559L);
     }
 
+    @Test
+    public void testFnv()
+    {
+        // https://nqv.github.io/fnv/
+        assertFunction("fnv1_32(from_hex(''))", BIGINT, 0x811c9dc5L + Integer.MIN_VALUE * 2L);
+        assertFunction("fnv1_32(from_hex('19'))", BIGINT, 0x050c5d06L);
+        assertFunction("fnv1_32(from_hex('F5'))", BIGINT, 0x050c5deaL); //Check for sign extension bug
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestVarbinaryFunctions.java]: Thanks for the nice test cases!

nit for the comment: space after `//`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestVarbinaryFunctions.java [hunk 6]
```diff
@@ -364,6 +364,29 @@ public void testCrc32()
         assertFunction("crc32(to_utf8('ABCDEFGHIJKLM'))", BIGINT, 4223167559L);
     }
 
+    @Test
+    public void testFnv()
+    {
+        // https://nqv.github.io/fnv/
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestVarbinaryFunctions.java]: nit: 
```
// ground truth result is generated via https://nqv.github.io/fnv/
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java (hunk 4 above)
Comment (widagdos) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: This is the only difference between fnv1 and fnv1a. Fnv1a is slightly better, so we can remove the fnv1. Or we can include both for completeness. what do you think?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java (hunk 2 above)
Comment (widagdos) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: It seems for sql functions, we use _ a lot, and only using lower case. Should we keep the old method names like ScalarFunction("fnv1_32")?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java (hunk 4 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/FnvHash.java]: @widagdos Sounds good. Let's include both for completeness . :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java [hunk 7]
```diff
@@ -315,6 +315,38 @@ public static long crc32(@SqlType(StandardTypes.VARBINARY) Slice slice)
         return crc32.getValue();
     }
 
+    @Description("compute fnv-1 32 bit")
+    @ScalarFunction("fnv1_32")
+    @SqlType(StandardTypes.BIGINT)
+    public static long fnv1Hash32(@SqlType(StandardTypes.VARBINARY) Slice slice)
+    {
+        return FnvHash.fnv1Hash32(slice);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: Please static import the functions per coding convention.

Comment (widagdos) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: So right now, the method names in VarbinaryFunctions are the same as method names in FnvHash. I think if we use static imports, by default, the code will call the locally defined functions, which will cause stack overflow due to the recursive call.

Should I rename the functions to be fnv1Hash32VarBinary(), etc?

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/VarbinaryFunctions.java]: Oh didn't realize that. This is fine then. Thanks!

**Comment by wenleix** [comment]
Merged #14139. Thanks for the contribution!

**Pull request merged**